File: python/src/aegis_trust/adapters/langchain.py
"""LangChain (Python) binder. Maps a ShieldedTool onto a LangChain tool.

The LangChain ``StructuredTool.from_function`` factory is INJECTED by the
caller, so aegis-trust takes no dependency on LangChain and is immune to its
internal version churn — only that public factory signature has to hold.
"""
from __future__ import annotations

from typing import Any, Callable

from aegis_trust.adapters.core import ShieldedTool


def to_langchain_tool(
    structured_tool_from_function: Callable[..., Any],
    tool: ShieldedTool,
    **extra: Any,
) -> Any:
    """Bind a :class:`ShieldedTool` to a LangChain (Python) StructuredTool.

    Example::

        from langchain_core.tools import StructuredTool
        from aegis_trust.adapters import shielded_tool, to_langchain_tool

        lookup = shielded_tool(
            name="customer_lookup",
            description="Look up a customer record by id for support.",
            purpose="customer_support",
            scope=["name", "issue"],
            handler=lambda customer_id: db.fetch(customer_id),
        )
        lc_tool = to_langchain_tool(StructuredTool.from_function, lookup)

    :param structured_tool_from_function: ``StructuredTool.from_function``
        (dependency-injected). The wrapped function returns the filtered value
        serialized to a string (what a LangChain tool emits to the model).
    :param tool: the shielded tool descriptor.
    """
    def _func(**kwargs: Any) -> str:
        return tool.call(**kwargs)

    kwargs: dict[str, Any] = {
        "func": _func,
        "name": tool.name,
        "description": tool.description,
    }
    if tool.schema is not None:
        kwargs["args_schema"] = tool.schema
    kwargs.update(extra)
    return structured_tool_from_function(**kwargs)

File: node/src/adapters/langchain.ts
// LangChain.js binder. Maps a ShieldedTool onto a LangChain StructuredTool.
//
// The LangChain `tool` factory is INJECTED by the caller rather than imported
// here, so aegis-trust takes no dependency on LangChain and is immune to its
// internal version churn — only the public `tool(func, config)` factory
// signature has to hold.

import type { ShieldedTool } from "./core.js";

/**
 * Minimal structural type of `@langchain/core/tools`' `tool` factory. The
 * wrapped function returns a string (LangChain tool outputs are text the model
 * reads), which is exactly what `ShieldedTool.call()` produces.
 */
export type LangChainToolFactory<R = unknown> = (
  func: (input: unknown) => Promise<string> | string,
  config: { name: string; description: string; schema?: unknown },
) => R;

/**
 * Bind a {@link ShieldedTool} to a LangChain.js StructuredTool.
 *
 * @example
 *   import { tool } from "@langchain/core/tools";
 *   import { shieldedTool, toLangChainTool } from "aegis-trust/adapters";
 *
 *   const lookup = shieldedTool({
 *     name: "customer_lookup",
 *     description: "Look up a customer record by id for support.",
 *     purpose: "customer_support",
 *     scope: ["name", "issue"],
 *     handler: async ({ id }: { id: string }) => db.fetch(id),
 *   });
 *
 *   const lcTool = toLangChainTool(tool, lookup);  // hand to AgentExecutor
 *
 * @param factory - the framework's `tool` factory (dependency-injected).
 * @param t - the shielded tool descriptor.
 * @returns whatever the injected `tool` factory returns (a StructuredTool) —
 *   the return type is inferred from the factory, so typed callers can pass the
 *   result straight to `AgentExecutor` without a cast.
 */
export function toLangChainTool<R>(factory: LangChainToolFactory<R>, t: ShieldedTool): R {
  return factory((input: unknown) => t.call(input), {
    name: t.name,
    description: t.description,
    ...(t.schema !== undefined ? { schema: t.schema } : {}),
  });
}
